File: lista_circular/lista.py
# Creamos la clase para el nodo
class Nodo:
    def __init__(self, dato=None):
        self.dato = dato
        self.siguiente = None

# Creamos la clase para la lista enlazada circular
class ListaCircular:
    def __init__(self):
        # El ultimo nodo de la lista
        self.ultimo = None

    # Metodo para insertar un nuevo nodo al inicio
    def insertar_al_inicio(self, dato):
        nuevo_nodo = Nodo(dato)
        
        # Si la lista esta vacia
        if self.ultimo is None:
            # el nuevo nodo se apunta a si mismo
            self.ultimo = nuevo_nodo
            nuevo_nodo.siguiente = self.ultimo
        else:
            # El nuevo nodo apunta a la cabeza actual
            nuevo_nodo.siguiente = self.ultimo.siguiente
            # El ultimo nodo apunta al nuevo nodo
            self.ultimo.siguiente = nuevo_nodo
    
    # Metodo para insertar un nuevo nodo al final
    def insertar_al_final(self, dato):
        nuevo_nodo = Nodo(dato)
        
        # Si la lista esta vacia
        if self.ultimo is None:
            # el nuevo nodo se apunta a si mismo
            self.ultimo = nuevo_nodo
            nuevo_nodo.siguiente = self.ultimo
        else:
            # El nuevo nodo apunta a la cabeza actual
            nuevo_nodo.siguiente = self.ultimo.siguiente
            # El ultimo nodo apunta al nuevo nodo
            self.ultimo.siguiente = nuevo_nodo
            # El ultimo ahora es el nuevo nodo
            self.ultimo = nuevo_nodo

    # Metodo para mostrar la lista
    def mostrar(self):
        if self.ultimo is None:
            print("La lista esta vacia.")
            return

        temp = self.ultimo.siguiente
        while True:
            print(temp.dato, end=" -> ")
            temp = temp.siguiente
            if temp == self.ultimo.siguiente:
                break
        print("...")

# Creamos una instancia de la lista y agregamos elementos
lista = ListaCircular()

lista.insertar_al_inicio(20)
lista.insertar_al_inicio(10)

print("La lista despues de insertar al inicio es:")
lista.mostrar()

lista.insertar_al_final(30)
lista.insertar_al_final(40)

print("\nLa lista despues de insertar al final es:")
lista.mostrar()

File: lista_circular/lista.cpp
#include <iostream>

// Definimos la estructura de un nodo
struct Nodo {
    int dato;
    Nodo* siguiente;
};

// Funcion para insertar un nuevo nodo al inicio de la lista circular
void insertarAlInicio(Nodo** ultimo, int nuevoDato) {
    // 1. Crear el nuevo nodo
    Nodo* nuevoNodo = new Nodo();
    nuevoNodo->dato = nuevoDato;
    
    // 2. Si la lista esta vacia
    if (*ultimo == NULL) {
        // el nuevo nodo se apunta a si mismo
        *ultimo = nuevoNodo;
        (*ultimo)->siguiente = *ultimo;
    } else {
        // 3. El nuevo nodo apunta a la cabeza actual
        nuevoNodo->siguiente = (*ultimo)->siguiente;
        // 4. El ultimo nodo apunta al nuevo nodo
        (*ultimo)->siguiente = nuevoNodo;
    }
}

// Funcion para insertar un nuevo nodo al final de la lista circular
void insertarAlFinal(Nodo** ultimo, int nuevoDato) {
    // 1. Crear el nuevo nodo
    Nodo* nuevoNodo = new Nodo();
    nuevoNodo->dato = nuevoDato;
    
    // 2. Si la lista esta vacia
    if (*ultimo == NULL) {
        // el nuevo nodo se apunta a si mismo
        *ultimo = nuevoNodo;
        (*ultimo)->siguiente = *ultimo;
    } else {
        // 3. El nuevo nodo apunta a la cabeza actual
        nuevoNodo->siguiente = (*ultimo)->siguiente;
        // 4. El ultimo nodo apunta al nuevo nodo
        (*ultimo)->siguiente = nuevoNodo;
        // 5. El ultimo ahora es el nuevo nodo
        *ultimo = nuevoNodo;
    }
}

// Funcion para mostrar la lista circular
void mostrarLista(Nodo* ultimo) {
    if (ultimo == NULL) {
        std::cout << "La lista esta vacia." << std::endl;
        return;
    }
    Nodo* temp = ultimo->siguiente;
    
    do {
        std::cout << temp->dato << " -> ";
        temp = temp->siguiente;
    } while (temp != ultimo->siguiente);
    
    std::cout << "..." << std::endl;
}

int main() {
    // Empezamos con una lista vacia
    Nodo* ultimo = NULL;

    // Insertamos 2 nodos al inicio
    insertarAlInicio(&ultimo, 20);
    insertarAlInicio(&ultimo, 10);

    std::cout << "La lista despues de insertar al inicio es: " << std::endl;
    mostrarLista(ultimo);
    
    // Insertamos 2 nodos al final
    insertarAlFinal(&ultimo, 30);
    insertarAlFinal(&ultimo, 40);

    std::cout << "\nLa lista despues de insertar al final es: " << std::endl;
    mostrarLista(ultimo);

    return 0;
}
